components/raftstore-v2/src/fsm/apply.rs

```diff
@@ -1,74 +1,101 @@
-            match self.receiver.try_recv() {
-                Ok(msg) => apply_task_buf.push(msg),
-                Err(e) => {
-                    if let TryRecvError::Disconnected = e {
-                        self.is_stopped = true;
-                    }
-                    return i - l;
-                }
-            }
-        }
-        apply_task_buf.capacity() - l
+impl<F: Fsm<Message = PeerMsg>, S: FsmScheduler<Fsm = F>> ApplyResReporter for Mailbox<F, S> {
+    fn report(&self, apply_res: ApplyRes) {
+        // TODO: check shutdown.
+        self.force_send(PeerMsg::ApplyRes(apply_res)).unwrap();
     }
 }
 
-impl<EK: KvEngine> Fsm for ApplyFsm<EK> {
-    type Message = ApplyTask;
+/// Schedule task to `ApplyFsm`.
+pub struct ApplyScheduler {
+    sender: Sender<ApplyTask>,
+}
 
+impl ApplyScheduler {
     #[inline]
-    fn is_stopped(&self) -> bool {
-        self.is_stopped
+    pub fn send(&self, task: ApplyTask) {
+        // TODO: ignore error when shutting down.
+        self.sender.send(task).unwrap();
     }
 }
 
-pub struct ApplyFsmDelegate<'a, EK: KvEngine> {
-    fsm: &'a mut ApplyFsm<EK>,
-    apply_ctx: &'a mut ApplyContext,
+pub struct ApplyFsm<EK: KvEngine, R> {
+    apply: Apply<EK, R>,
+    receiver: Receiver<ApplyTask>,
 }
 
-impl<'a, EK: KvEngine> ApplyFsmDelegate<'a, EK> {
-    pub fn new(fsm: &'a mut ApplyFsm<EK>, apply_ctx: &'a mut ApplyContext) -> Self {
-        Self { fsm, apply_ctx }
+impl<EK: KvEngine, R> ApplyFsm<EK, R> {
+    pub fn new(
+        region_state: RegionLocalState,
+        res_reporter: R,
+        remote_tablet: CachedTablet<EK>,
+        logger: Logger,
+    ) -> (ApplyScheduler, Self) {
+        let (tx, rx) = future::unbounded(WakePolicy::Immediately);
+        let apply = Apply::new(region_state, res_reporter, remote_tablet, logger);
+        (
+            ApplyScheduler { sender: tx },
+            Self {
+                apply,
+                receiver: rx,
+            },
+        )
     }
+}
 
-    pub fn handle_msgs(&self, apply_task_buf: &mut Vec<ApplyTask>) {
-        for task in apply_task_buf.drain(..) {
-            // TODO: handle the tasks.
+impl<EK: KvEngine, R: ApplyResReporter> ApplyFsm<EK, R> {
+    pub async fn handle_all_tasks(&mut self) {
+        loop {
+            let mut task = match self.receiver.next().await {
+                Some(t) => t,
+                None => return,
+            };
+            loop {
+                match task {
+                    ApplyTask::CommittedEntries(ce) => self.apply.apply_committed_entries(ce).await,
+                }
+
+                // TODO: yield after some time.
+
+                // Perhaps spin sometime?
+                match self.receiver.try_recv() {
+                    Ok(t) => task = t,
+                    Err(TryRecvError::Empty) => break,
+                    Err(TryRecvError::Disconnected) => return,
+                }
+            }
+            self.apply.flush();
```

Comment:
should it call flush in case there're too many entries?

Reply:
Yes, it should. It's just not implemented.

Reply:
Then please add a TODO here.

Reply:
Also I think maybe we should flush not by fixed batch (entry) size, but by actual bytes size, it should be more accurate for time to spent.